Improve the CI speed and accuracy

Elv13 opened the pull request:
![image](https://cloud.githubusercontent.com/assets/340384/20650845/9b59b058-b4a6-11e6-805c-534277aa7565.png)

Fix #1142 
Fix #982 
Fix #962
Fix #964 
Close #1238

Elv13 commented:
@psychon codecov somehow fails, but looking at coverall, this PR makes the build ~4 minutes faster. Of course, its ugly. On the other hand, I think this fix the coverage line delta, but I can't be sure until codecov runs again.

psychon commented:
Well, the message is:
```
Could not load stats file luacov.stats.out.
Run your Lua program with -lluacov and then rerun luacov.
```
It's not codecov that fails, but luacov.

coveralls seems to have the same problem: Coverage data from `tests/examples` is missing. This is just not a fatal error for coveralls.

I think you just rediscovered the problem that https://github.com/awesomeWM/awesome/pull/964 / https://github.com/awesomeWM/awesome/issues/982 already found.

Elv13 commented:
/me Down the rabbit hole I went, many issues I found, many issues I fixed, many issues remains.

@psychon Your patch to stop the screens for leaking is totally working. It also prevent the tags from leaking. In returns, on my local computer, it does fully fix the exponential memory usage. It doesn't fix it when the coverage is enabled. Therefor I blame luacov for it. The more code is executed, the more memory luacov takes. I propose to close both slow CI related issues as they are not caused by Awesome.

Also, coveralls isn't happy at all to lose the test example coverage. CodeCov don't care much because of some miracle circumstances where the number of lines removed is exactly proportional to the number of misses remove due to the coverage being correct (WTF). The CodeCov lines are now correct, the offset caused by the example tests is now corrected. I added an attempted fix for Coveralls, but it doesn't help, so I leave this to someone who knows more about it.

* The coverage no longer include the tests
* The build is 3.5-4 minutes faster by trimming some fat where we can
* The screens and tags are no longer leaking in multi screen tests
* The coverage data no longer have an offset due to the CMake `configure_file` insertions
* Work has been removed from the slower builds and added to the faster ones
* The investigation why 2 of the builds are very slow is completed
* Open issues have been fixed

Other possible speed holes:

 * Disable the wibar on the client multi-screen tests, it should trim 45s
 * Install less dependencies TBD
 * Don't install/use ldoc on the slower targets (save the install time of penlight and ldoc, save 45s)
 * Create our own debian repository and stop pulling from git and luarock all the time (save 1.5 minute)
 * Explore using Docker.io images on Travis and see its smart enough to cache the first few build steps (replace the 3 previous bullet points, save 3 minutes)

blueyed commented:
What needs to be done here?
"coverage/coveralls" complains about coverage loss: -8.3%.

psychon commented:
Should I take a look before the release? I thought the priorities were different.

blueyed commented:
I don't think it is a release blocker - I came across it when waiting for Travis, and remembered that this is supposed to make it faster.

blueyed commented on .travis.yml:
```diff
@@ -89,8 +89,11 @@ install:
-  - travis_retry sudo luarocks install ldoc 1.4.4
-  - travis_retry sudo luarocks install lua-discount
+  - |
+    if [ "$DO_COVERAGE" == "0" || "$DO_COVERAGE" == "" ]; then
```
Should check for `$BUILD_APIDOC` here instead?!

blueyed commented on awesomeConfig.cmake:
```diff
@@ -77,7 +77,7 @@ if(GENERATE_MANPAGES)
 
 if(GENERATE_DOC)
-    if(NOT LDOC_EXECUTABLE)
+    if(NOT LDOC_EXECUTABLE OR DO_COVERAGE)
```
Should not check for DO_COVERAGE here and above: the build should (and does) not install ldoc instead.

psychon replied:
So... this can be removed?

blueyed replied:
This part was changed by me, too.
Should amend the previous commit.

blueyed commented on .travis.yml:
```diff
@@ -149,16 +156,16 @@ script:
         travis_fold_start "DO_COVERAGE"
+
+        (mkdir -p tests/examples/build && cd tests/examples/build && cmake $CMAKE_ARGS -DUSE_LCOV=1 "..")
+        do_codecov samples tests/examples/build
```
Maybe there is a problem with relative paths (in `luacov.stats.out`, which gets copied).
What about `cd`ing here instead?!

blueyed commented on .travis.yml:
```diff
@@ -10,9 +10,9 @@ branches:
+    - LUA=5.2 DO_CHECKQA=1 LUANAME=lua5.2 DO_COVERAGE=coveralls
     # luajit: installed from source.
-    - LUA=5.1 LUANAME=luajit-2.0 LUALIBRARY=/usr/lib/libluajit-5.1.so LUAROCKS_ARGS=--lua-suffix=jit-2.0.4 TEST_PREV_COMMITS=1
+    - LUA=5.1 LUANAME=luajit-2.0 LUALIBRARY=/usr/lib/libluajit-5.1.so BUILD_APIDOC=true LUAROCKS_ARGS=--lua-suffix=jit-2.0.4 TEST_PREV_COMMITS=1
```
Might easily be the slowest though, because of TEST_PREV_COMMITS.

psychon replied:
Then I propose to drop `TEST_PREV_COMMITS`. The idea is nice, but its slow down is a problem for larger pull requests.

blueyed replied:
Seems to be not working as expected currently anyway: looks like it tests only one of them.

> its slow down is a problem for larger pull requests.

Well, I think that is OK, given the benefit of ensuring that git-bisect works (better).
And given that it uses the existing build (cache), it is probably not so slow after all.

My point in the comment above was just that this should be taken into account when shuffling the build matrix.

blueyed commented on awesomeConfig.cmake:
```diff
@@ -329,19 +328,25 @@ if(GENERATE_DOC)
 endif()
 
 # {{{ Configure files
```
Changes here are hard to understand.
Should have some comment(s), and I dislike using "DO_COVERAGE" here, but it cannot be avoided probably.

blueyed commented on awesomeConfig.cmake:
```diff
@@ -352,6 +357,27 @@ foreach(file ${AWESOME_CONFIGURE_FILES})
+    ${awesome_lua_configure_files}
+)
+
+if(NOT DO_COVERAGE)
```
Switch the `if` blocks to have the positive one first.
Also a comment here would help.

blueyed commented on awesomeConfig.cmake:
```diff
@@ -352,6 +357,27 @@ foreach(file ${AWESOME_CONFIGURE_FILES})
                    @ONLY)
 endforeach()
+
+set(AWESOME_CONFIGURE_NO_COV_FILES
```
Better name: `AWESOME_CONFIGURE_COPY_ONLY_FILES`?

blueyed commented on awesomeConfig.cmake:
```diff
@@ -376,6 +376,12 @@ else()
                     COPYONLY)
     endforeach()
+
+    # There is no way to avoid that one
```
What for / why?
Could not find "@@" in there.

And if required, this part could be moved into a separate file maybe?!

blueyed commented on awesomeConfig.cmake:
```diff
@@ -352,6 +357,33 @@ foreach(file ${AWESOME_CONFIGURE_FILES})
+    configure_file(${SOURCE_DIR}/lib/awful/util.lua
+                    ${BUILD_DIR}/lib/awful/util.lua
+                    ESCAPE_QUOTES
+                    @ONLY)
```
What is it for / why can't it be avoided?
Could not find "@@" in there.

And if required, this part could be moved into a separate file maybe?!

blueyed replied:
Ah:
```lua
--- Get the path to the directory where themes are installed.
-- @return A string with the requested path with a slash at the end.
function util.get_themes_dir()
    return "@AWESOME_THEMES_PATH@" .. "/"
end

--- Get the path to the directory where our icons are installed.
-- @return A string with the requested path with a slash at the end.
function util.get_awesome_icon_dir()
    return "@AWESOME_ICON_PATH@" .. "/"
end
```

Should/can we move this to a separate file?
Or is it not too much of a hack to handle the whole of `util.lua` in a special way here?

blueyed commented:
As per the discussion in https://github.com/awesomeWM/awesome/pull/1238#issuecomment-269131480 we might want to remove coveralls (or ignore it for now?!).
But I've left a comment above which might explain it (and we could `cd` instead).
Will look into it later.

Elv13 commented:
Thanks for taking a look. I guess you now know why I put that image at the top. Fixing this has been only frustration for @psychon and me for a while now... Do you think Coveralls is worth all that trouble? I got it to work fine with CodeCov. IMO your time is better spent elsewhere than trying yet more changes. My own commits were already rebased a hundred time to get to this point.

blueyed commented:
I'll poke a bit more.
It seems like the current luacov failed to work with codecov also (that's why it was pinned): it was not generating coverage output for the functional tests: https://travis-ci.org/awesomeWM/awesome/jobs/186857911#L4089
9abe68c hopefully fixes this.

I agree that we do not really need coveralls also, but it's very nice to compare the coverage from two different tools / processors though, to see if something goes wrong.

blueyed commented:
@Elv13 @psychon 
It worked out finally..
Will clean it up a bit more now.

blueyed commented on CMakeLists.txt:
```diff
@@ -388,17 +388,16 @@ if(BUSTED_EXECUTABLE)
-        "BUILD_DIRECTORY=${CMAKE_BINARY_DIR}/"
         ${BUSTED_EXECUTABLE} "--helper=${CMAKE_SOURCE_DIR}/spec/preload.lua"
-        "--lpath=${CMAKE_BINARY_DIR}/lib/?.lua;${CMAKE_BINARY_DIR}/lib/?/init.lua;spec/?.lua"
+        "--lpath=lib/?.lua;lib/?/init.lua;spec/?.lua"
```
This might be wrong, but is the easiest way to get the coverage from there.
It can be done by setting `SOURCE_DIRECTORY` though instead (probably), or rather by adjusting `.luacov` for it.

blueyed commented on .luacov:
```diff
@@ -20,12 +20,20 @@ return {
         "^awesomerc.lua",
+
+        -- Include awesomerc.lua (in build dir with integration tests).
+        escape_pattern(build) .. "/awesomerc.lua",
```
This might not be required when using a relative path for the RC file (awesome gets run from the build dir already)?!

blueyed replied:
But `awesomerc.lua` is where the `luacov.runner` gets added, so that might be irrelevant..

blueyed commented:
awesomewmrc.lua gets not included.
I will remove it from this PR for now.

Elv13 commented:
👍

psychon commented on tests/examples/CMakeLists.txt:
```diff
@@ -14,22 +14,44 @@ cmake_minimum_required(VERSION 3.0.0)
+
+    if (NOT EXISTS "${LCOV_PATH}")
+        message(WARNING "The luacov configuration file was not found")
+        unset(LCOV_PATH)
```
Why a warning instead of an error?

blueyed replied:
This part has been removed by me later.. should amend the previous commit here..

psychon commented on tests/examples/CMakeLists.txt:
```diff
@@ -14,22 +14,44 @@ cmake_minimum_required(VERSION 3.0.0)
+    set(ENV{SOURCE_DIRECTORY} ${SOURCE_DIR})
+else()
+    set(SOURCE_DIR ${CMAKE_SOURCE_DIR})
+endif()
```
"if `SOURCE_DIR` is set, overwrite it". That seems odd to me. How about giving this variable a different name instead (dunno, `TOP_SOURCE_DIR`)?

blueyed replied:
It is not meant to be overwritten, I guess.
But another name makes sense!

psychon commented on tests/examples/CMakeLists.txt:
```diff
@@ -14,22 +14,44 @@ cmake_minimum_required(VERSION 3.0.0)
 # Get and update the LUA_PATH so the scripts can be executed without Awesome.
 execute_process(COMMAND lua -e print\(package.path\) OUTPUT_VARIABLE "LUA_PATH_")
 
```
For the commit message: How about some more details. The commit message does not even say that this commit fixes running the doc tests without compiling. I'd like to know what exactly was broken and was fixed (and perhaps also that "running the doc tests without compiling" means running `cmake` in the `tests/examples` directory (does it mean that?))

psychon commented on awesomeConfig.cmake:
```diff
@@ -355,6 +360,33 @@ foreach(file ${AWESOME_CONFIGURE_FILES})
+                    @ONLY)
+    endforeach()
+endif()
+
```
Sigh. More bad hacks and band-aids. My long-term plan is still to change things so that bug reports with line numbers are useful again, but for now if this hack makes you happy...

blueyed replied:
In this case line numbers should match the source?!

psychon replied:
The line numbers that the coverage tool sees, yes. But line numbers from bug reports will still be different from what we have in Git.

blueyed replied:
Ah, of course.
I agree that we should get back to that state in the long run.

psychon commented on .travis.yml:
```diff
@@ -147,16 +151,16 @@ script:
+        (mkdir -p tests/examples/build && cd tests/examples/build && cmake $CMAKE_ARGS -DUSE_LCOV=1 "..")
+        do_codecov samples tests/examples/build
+
         (make check-unit-coverage \
```
Could someone please explain this commit? For example in the commit message? I don't like the copy of `luacov.stats.out` (why??), I don't like most of it.

psychon commented on .travis.yml:
```diff
@@ -89,8 +89,11 @@ install:
+    if [ "$BUILD_APIDOC" == "true" ]; then
+      travis_retry sudo luarocks install ldoc 1.4.4
+      travis_retry sudo luarocks install lua-discount
+    fi
```
As Daniel already pointed out: Doesn't this make changes in the CMake files from earlier commits obsolete?

psychon commented on .luacov:
```diff
@@ -1,33 +1,29 @@
-local source = os.getenv("SOURCE_DIRECTORY")
+local source = os.getenv("SOURCE_DIRECTORY") or os.getenv('PWD')
+local build = os.getenv("BUILD_DIRECTORY") or (source .. "/build")
+local lib_dir = os.getenv("AWESOME_LIB_DIR") or (source .. "/lib")
```
According to the commit message, this should be more than one commit. I have a hard time following all of this. Why the switch to a different version of luacov? Should that perhaps just be unpinned instead? What does `set -v` do and what is its connection with this PR?

blueyed replied:
Yeah, I ended up squashing everything.
And there are reverts for previous commits (from this PR) in there..
I could look into cleaning that up again.

`set -v` prints the executed commands, but less verbose than `-x`. Travis seems to do that already / and/or it annoyed me.

blueyed replied:
Do you mean ldoc instead of luacov?

blueyed replied:
Oh..

>      - Use latest luacov (0.12.0-1) again                                                                            
>       This reverts parts of 4cc6a815.                                                    
>       I think it is better to fix any failure that 4cc6a815 tried to work around.

psychon commented on .travis.yml:
```diff
@@ -92,7 +92,7 @@ install:
   - |
     if [ "$BUILD_APIDOC" == "true" ]; then
-      travis_retry sudo luarocks install ldoc 1.4.4
+      travis_retry sudo luarocks install ldoc 1.4.6
```
How does this help with speed and accuracy? Why is ldoc still pinned to a specific version?

blueyed replied:
It was pinned for some workaround - mentioned in the commit message.
It can be upgraded now, but is still pinned to avoid unexpected changes.

blueyed replied:
I've thought of luacov with my previous comment.
Pulled out the update of ldoc into https://github.com/awesomeWM/awesome/pull/1334.

blueyed commented:
@psychon 
Thanks for the review.
You might want to look at the overall diff again for now, and not the individual commits.

blueyed commented on tests/_multi_screen.lua:
```diff
@@ -411,6 +411,7 @@ local function add_steps(real_steps, new_steps)
         -- Cleanup
         table.insert(real_steps, function()
+            collectgarbage("collect")
             if #client.get() == 0 then return true end
```
Pulled out here, and changed accordingly: https://github.com/awesomeWM/awesome/pull/1333.

blueyed commented:
Updated/cleaned it a bit.

I suggest squashing the commits from @Elv13 into my big one.

Elv13 commented:
> I suggest squashing the commits from @Elv13 into my big one.

Do as you wish

Elv13 commented:
Ping? Also, I don't mind if you just redo the commits and remove me as the author (and merge your changes with mines) so the commits are nicer. I also don't mind if this is squashed into a giant commit.

psychon commented on tests/run.sh:
```diff
@@ -23,7 +23,7 @@ fi
 cd -P -- "$(dirname -- "$0")"
 this_dir=$PWD
-source_dir="$PWD/.."
+source_dir="${this_dir%/*}"
```
This assumes that `$this_dir` (thus `$PWD`) contains a slash (and it was not all that easy to figure out what this does). I'm not really sure if this always works correctly...?

blueyed replied:
Yes, I think "$PWD" will always be absolute, and therefore contain a slash.

psychon commented on tests/examples/CMakeLists.txt:
```diff
@@ -14,20 +14,45 @@ cmake_minimum_required(VERSION 3.0.0)
+    # Used by .luacov.
+    set(ENV{SOURCE_DIRECTORY} ${SOURCE_DIR})
+else()
+    set(SOURCE_DIR ${CMAKE_SOURCE_DIR})
```
Still overwrites `$SOURCE_DIR`

psychon commented on tests/examples/CMakeLists.txt:
```diff
@@ -14,20 +14,45 @@ cmake_minimum_required(VERSION 3.0.0)
+    message("${TEST_ERROR}")
+    set(ENV{DO_COVERAGE} 0)
+  endif()
+endif()
```
....why? I would just let it fail instead instead of semi-silently just not generating coverage data.

psychon commented on awesomeConfig.cmake:
```diff
@@ -332,19 +335,25 @@ if(GENERATE_DOC)
-    ${SOURCE_DIR}/*.h
+    ${SOURCE_DIR}/*.h)
+
+file(GLOB awesome_c_configure_files RELATIVE ${SOURCE_DIR}
```
Why split this up in two? We do not generate coverage information for the C code anyway (should we?).

blueyed replied:
Yes, we should eventually.. currently it makes no sense to split them, but it's also not a problem really.  I'll leave it as-is for now.

psychon commented:
Codecov currently [says](https://github.com/awesomeWM/awesome/pull/1239#issuecomment-267736185):
```
@@             master      #1239   diff @@
==========================================
  Files           242        116    -126   
  Lines         16511      12107   -4404   
  Methods           0          0           
  Messages          0          0           
  Branches          0          0           
==========================================
- Hits          12072       8837   -3235   
+ Misses         4439       3270   -1169   
  Partials          0          0           
```
Why do 126 files go missing? (And I guess someone will tell me the 4404 files are documentation, but even including the DOC_HIDE lines, templates and CMake files, we only have 3955 such lines according to `cat $(find tests/examples -type f) | wc -l`).

blueyed commented:
About the missing files: in master, `tests` and `themes/xresources/assets.lua` is also included: https://codecov.io/gh/awesomeWM/awesome/branch/master.

I will bring them back.. we should probably just include everything from our build and source dirs.  Would fix https://github.com/awesomeWM/awesome/issues/1315 then also.

blueyed commented:
FWIW, there was some timeout just now with the coveralls build:
```
== Running /home/travis/build/awesomeWM/awesome/tests/test-awful-client.lua ==
2017-01-03 16:41:24 W: awesome: a_glib_poll:389: Last main loop iteration took 1.272816 seconds! Increasing limit for this warning to that value.
2017-01-03 16:41:28 W: awesome: a_glib_poll:389: Last main loop iteration took 3.037822 seconds! Increasing limit for this warning to that value.
2017-01-03 16:41:44 W: awesome: a_glib_poll:389: Last main loop iteration took 3.144219 seconds! Increasing limit for this warning to that value.
2017-01-03 16:41:57 W: awesome: a_glib_poll:389: Last main loop iteration took 3.237561 seconds! Increasing limit for this warning to that value.
2017-01-03 16:42:35 W: awesome: a_glib_poll:389: Last main loop iteration took 4.407701 seconds! Increasing limit for this warning to that value.
2017-01-03 16:43:12 W: awesome: a_glib_poll:389: Last main loop iteration took 6.451608 seconds! Increasing limit for this warning to that value.
2017-01-03 16:44:07 W: awesome: a_glib_poll:389: Last main loop iteration took 7.415468 seconds! Increasing limit for this warning to that value.
Awesome was killed due to timeout after 180 seconds
===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-awful-client.lua <===
```

Elv13 commented:
> FWIW, there was some timeout just now with the coveralls build:

They happen from time to time. This PR doesn't touch the integration tests, so it probably isn't to blame.

blueyed commented:
Hmm, luacov did not update its comment?!

Elv13 commented:
> Why do 126 files go missing?

@psychon #964

psychon commented:
So... we make @blueyed clean up the commits and then this gets merged?

Elv13 commented:
Yes, once the regression from the last fixup is fixed. Namely only 2 of the 5 targets remains ( see https://travis-ci.org/awesomeWM/awesome/builds/188582885 ).

Elv13 commented:
@blueyed Is it possible that your last changes made it show coverage for the tests again? The -126 files is gone from the new coverage. This was done on purpose to fix #964.

blueyed commented:
The TEST commit will be removed, of course.
What is bad about covering everything, especially the tests itself?
I wonder about things like https://codecov.io/gh/awesomeWM/awesome/src/8af82ef7fd1430b701eaf1fbc49e317070871305/tests/examples/shims/mouse.lua though, which should be covered 100% shouldn't it?

Elv13 commented:
> What is bad about covering everything, especially the tests itself?

I personally don't mind. @psychon said otherwise at some point. I also agree that the coverage percentage with the tests is a little inflated and misrepresent the reality. As for the link you provided, the shims are not expected to have full coverage. They mimic CAPI and are fixed only when a function is required. Some other documentation examples (such as the prompt ones) are also not fully covered and are not intended to be[1].

[1] It would be cool, but is impractical and not worth the effort.

blueyed commented:
@Elv13 
> As for the link you provided, the shims are not expected to have full coverage. They mimic CAPI and are fixed only when a function is required.

But shouldn't line 1 there be covered (`local screen = require("screen")`)?!
Also only the function body is covered, but not the function definition..
(is this really expected like that)

blueyed commented:
Similar: https://codecov.io/gh/awesomeWM/awesome/src/8af82ef7fd1430b701eaf1fbc49e317070871305/tests/examples/shims/_common_template.lua - if I add some `error()` there, it gets executed with `DO_COVERAGE=1 cmake tests/examples` certainly.
So it looks like an issue with `luacov` after all?!

I am getting the following in `luacov.stats.out`:

> 24:/home/user/src/awesome/tests/examples/shims/_common_template.lua
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 153 151

Elv13 commented on .travis.yml:
```diff
@@ -161,7 +161,7 @@ script:
         (make check-unit-coverage \
          && do_codecov unittests \
-         && sed "1 s~^~require('luacov.runner')('$PWD/.luacov'); \0~" build/awesomerc.lua \
+         && sed -i "1 s~^~require('luacov.runner')('$PWD/.luacov'); \0~" build/awesomerc.lua \
```
The "original line numbers" would require to use `../awesomerc.lua`. `build/awesomerc.lua` have been modified by the build system to remove the section markers used when generating the documentation.

blueyed replied:
Ok.  I will change tests/run.sh to handle this (instead of here), which would allow for `DO_COVERAGE=1 make check` to handle it there.

blueyed commented:
Looking good by now again.

Codecov has 252 files covered (only `preload.lua` in `spec/`, because `luacov` could not `stat` the relative `spec/` files).  This should be fixed in the commit I've just pushed.
Coveralls has 268 files covered (16 additional files in `spec/`).

blueyed commented:
Rebased.
Coverage increased to 78.99% on both.
I'll merge this tonight probably.

Elv13 commented:
It's still a bit sad that all the speedups went away. The first build after both coveralls and codecov was fixed was in the 8.x minutes range. Now, there is "5% more coverage gamification tokens", but its 11m long. Just as the coverage for the C code, I will have to ask: is there real value from having the test covered. Is it worth waiting 2:30m more for each builds? @psychon, what do you think about this?

I am not saying this PR isn't a great job: it is. Thanks a lot for it, I know how long it took you to fix all the corner case my version had. It's just IMO that the last few changes went in favor of the "accuracy" part but did hurt the "speed" part a lot.

blueyed commented:
Well, it is only reporting more coverage, i.e. not excluding it - the example tests are/should run with coverage, and therefore also the tests are covered.
AFAIK once coverage is enabled, it will be tracked across all files, doesn't it?

The initial 8m runs seem to not have generated coverage after all?!
I could try another build with adding the excludes for tests/ again, but this probably not what you've meant?!

Elv13 commented:
> The initial 8m runs seem to not have generated coverage after all?!

Codecov was correctly generated (with only the `lib/` file, that was done on purpose) and took 8:35m. When you initially fixed coveralls, it was in the same ballpark too.

> AFAIK once coverage is enabled, it will be tracked across all files, doesn't it?

I see the logic in that, but if codecov was getting the correct (as in, the same coverage numbers as coveralls in other unrelated PRs)percentage in 8:35m (average over 5 runs), then why is it always at 9:45m (+/- 30 sec) right now? This isn't due to bad luck with the Travis nodes, not this many time in a row.

If by some luck going back to 188 (at 73%) covered files put it back below 9 minutes, then I argue it is worth more than having 268 files covered (the tests & shims) and a shiny 79% coverage badge.

Elv13 commented:
Run1: coveralls 9:20, codecov 8:57

(I will update this post)

blueyed commented on .travis.yml:
```diff
@@ -226,6 +225,5 @@ after_success:
   - |
     set -e
     if [ "$DO_COVERAGE" = "coveralls" ]; then
-      cat build/luacov.stats.out
```
Will change it to `test -f`, since this is where missing coverage info used to bail out.

blueyed commented:
@Elv13 
Thanks.
What have you tried?  Changing the `include` in `.luacov`?

blueyed commented:
@Elv13
I've tried `time (DO_COVERAGE=1 cmake ..)` in `tests/examples/build` with and without including lib/ and awesomerc.lua.  And it was actually a bit slower when not including everything (although it should have less (disk) cache misses):

Full (1st)
43.24s user 0.57s system 99% cpu 43.995 total

Only lib/ + awesomerc.lua (2nd)
45.52s user 0.63s system 99% cpu 46.373 total

I think we should address performance/speed of the builds in a new PR, based on the suggestions you've provided already, e.g. by using (pre-built) Docker base images, which have all the deps installed already.

blueyed commented:
🎆
